crates/vim/src/replace.rs

```diff
@@ -0,0 +1,174 @@
+use std::sync::Arc;
+
+use crate::{
+    motion::Motion,
+    state::{Mode, Operator},
+    Vim,
+};
+use editor::{movement, Bias};
+use gpui::{actions, ViewContext, WindowContext};
+use log::error;
+use workspace::Workspace;
+
+actions!(vim, [ToggleReplace]);
+
+pub fn register(workspace: &mut Workspace, _: &mut ViewContext<Workspace>) {
+    workspace.register_action(|_, _: &ToggleReplace, cx: &mut ViewContext<Workspace>| {
+        Vim::update(cx, |vim, cx| {
+            if vim.state().mode == Mode::Replace {
+                vim.switch_mode(Mode::Normal, false, cx);
+            } else {
+                vim.switch_mode(Mode::Replace, false, cx);
+                vim.update_active_editor(cx, |_, editor, cx| {
+                    editor.set_last_snapshot(Some(editor.buffer().clone().read(cx).snapshot(cx)));
+                });
+            }
+        });
+    });
+}
+
+pub fn replace_motion(
+    motion: Motion,
+    operator: Option<Operator>,
+    times: Option<usize>,
+    cx: &mut WindowContext,
+) {
+    Vim::update(cx, |vim, cx| {
+        match operator {
+            None => undo_replace_motion(vim, motion, times, cx),
+            Some(operator) => {
+                // Can't do anything for text objects, Ignoring
+                error!("Unexpected replace mode motion operator: {:?}", operator)
+            }
+        }
+    });
+}
+
+pub(crate) fn multi_replace(text: Arc<str>, cx: &mut WindowContext) {
+    Vim::update(cx, |vim, cx| {
+        vim.stop_recording();
```

Comment:
You don't want to stop recording until someone leaves Replace mode. That way `.` should redo the full replacement. (though as I said above, I'm happy if that doesn't quite work for v1)